scripts/genVis.py

```diff
@@ -1,58 +1,140 @@
+    DRDF['umap1'] = umap_embed[:, 0]
+    DRDF['umap2'] = umap_embed[:, 1]
+    if umap_dim == 3:
+        DRDF['umap3'] = umap_embed[:, 2]
 
-(opts, args) = parser.parse_args()
-inFile = opts.inFile
-tsne_flag = opts.tsne
 
+# Prepare time-dependent color scheme
+ptDF = pd.read_csv(timeFile, sep=',', index_col=0)
+color_scale = max(ptDF["Time"])
+colors_raw = [h.split('_')[1] for h in DF.columns]
+colors_raw = [int(i) for i in colors_raw]
+colors = [x / color_scale for x in colors_raw]
+experiments = set([h.split('_')[0] for h in DF.columns])
+DRDF['time'] = colors
 
+# Prepare cluster-dependent color scheme
+if args.handleExternalClusterFile:
+    CF = pd.read_csv(clusterFile, sep=',', index_col=0)
+    cluster_colors_raw = CF['cl'].tolist()
+    cluster_color_scale = max(CF['cl'])
+    cluster_colors = [y / color_scale for y in cluster_colors_raw]
+else:
+    cluster_colors = list(itertools.repeat(.5, len(DF.columns)))
 
+# Write dimensionality reduction data to text file
+DRDF.to_csv(inFile + '_dimred.txt')
 
-DF = pd.read_csv(inFile,sep=',',index_col=0)
-Cells = DF.T.values
 
-####################
-# Do PCA and tSNE
-PC = PCA(n_components=2).fit_transform(Cells)
-embed = TSNE(n_components=2).fit_transform(Cells)
-####################    
-ptDF = pd.read_csv(opts.pseudoTimeFile, sep=',', index_col=0)
+# Common method for plotting
+def make_subplot(axis1, axis1label, axis2, axis2label, axis3, axis3label, dim):
+    plot_title = ' '.join(args.plotTitle)
+    f, ax = plt.subplots(1, 2, figsize=(10, 5))
+    plt.rcParams['image.cmap'] = 'viridis'
+    if dim == 3:
+        ax[0].set_axis_off()
+        ax[0] = f.add_subplot(1, 2, 1, projection="3d")
+        ax[0].scatter3D(DRDF[axis1], DRDF[axis2], DRDF[axis3], c=DRDF['time'])
+        ax[0].set_zlabel(axis3label)
+    else:
+        ax[0].scatter(DRDF[axis1], DRDF[axis2], c=DRDF['time'])
+    ax[0].set_xlabel(axis1label)
+    ax[0].set_ylabel(axis2label)
+    ax[0].set_aspect('auto')
+    ax[0].set_title('Simulation Time')
+
+    plt.rcParams['image.cmap'] = 'Spectral'
+    if dim == 3:
+        ax[1].set_axis_off()
+        ax[1] = f.add_subplot(1, 2, 2, projection="3d")
+        ax[1].scatter3D(DRDF[axis1], DRDF[axis2], DRDF[axis3], c=cluster_colors)
+        ax[1].set_zlabel(axis3label)
+    else:
+        ax[1].scatter(DRDF[axis1], DRDF[axis2], c=cluster_colors)
+    ax[1].set_xlabel(axis1label)
+    ax[1].set_ylabel(axis2label)
+    ax[1].set_aspect('auto')
+    ax[1].set_title('Clusters')
+
+    plt.suptitle(plot_title, fontsize=20)
 
-colors = ptDF.min(axis='columns').values
-print(colors)
-experiments = set([h.split('_')[0] for h in DF.columns])
-PCDF = pd.DataFrame(PC,columns=['PC1','PC2'],index=pd.Index(list(DF.columns)))
 
-PCDF['tsne1'] = embed[:,0]
-PCDF['tsne2'] = embed[:,1]
-    
-PCDF['time'] = colors
-PCDF.to_csv(inFile + '_dimred.txt')
+# t-SNE plotting
+if tsne_flag:
+    make_subplot('tsne1', 't-SNE 1', 'tsne2', 't-SNE 2', 'tsne3', 't-SNE 3', tsne_dim)
```

Comment:
Explain in a comment that make_subtplot will not use the fifth and sixth arguments in the plot if tsne_dim is 2.

Reply:
As I was about to write this comment, you have brought up a valid point. I should not have to pass this many arguments...I can hear my introductory CS TA scolding me for not using a for loop. You will see fewer arguments required to be passed for the `make_subplot` method in the next commit.